### PR TITLE
Update documentation for advanced tensorflow usage

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- Closes https://github.com/RasaHQ/rasa/issues/5347 and address review comments from https://github.com/RasaHQ/rasa/pull/5266
- Add more details and suggestions on which options to use when.

I don't think linking to more official tensorflow docs is a good idea because they are currently very lean and confusing themselves. In terms of information, there isn't anything more that people can extract out of it from a usage with Rasa perspective. For example, this guide on using a gpu https://www.tensorflow.org/guide/gpu is completely tensorflow focussed.  
API documentation for methods around CPU parallelism are very skinny - https://www.tensorflow.org/api_docs/python/tf/config/threading/set_intra_op_parallelism_threads and https://www.tensorflow.org/api_docs/python/tf/config/threading/set_inter_op_parallelism_threads

**Status (please check what you already did)**:
- [x] updated the documentation

### CODE REVIEW

**Comment by akelad** [comment]
@dakshvar22 before I continue reviewing this 

"These options are useful if you have implemented a custom component as part of your pipeline" - is this docs page even relevant to the components we have currently implemented in our repo then?

**Comment by dakshvar22** [comment]
@akelad The options will apply to all existing components as well as custom components that are run as part of the Rasa OS process.
I wanted to add this line because users can implement a custom component comprising of a model which they have built themselves and hence they are in a better position to adjust values for these parameters.
For users who are advanced tensorflow users and understand the architectures of our models like DIET can also take a call on these options if they wish to.

File: docs/api/tensorflow_usage.rst [hunk 1]
```diff
@@ -14,19 +14,32 @@ All configuration options are specified using environment variables as shown in
 Optimizing CPU Performance
 --------------------------
 
+.. note::
+    It is advisable to configure these options only if you are an advanced tensorflow user and understand
+    how different operations are implemented under the hood.
+
+These options are useful if you have implemented a custom component as part of your pipeline. Based on the tensorflow operations
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]: let's rephrase this to clarify that you should know the workings of an implementation well to know whether these configurations make sense - whether that's their own custom implementation or whether they've figured out our implementation

File: docs/api/tensorflow_usage.rst [hunk 2]
```diff
@@ -14,19 +14,32 @@ All configuration options are specified using environment variables as shown in
...
+that the component uses, you can leverage multi-core CPU parallelism by tuning these options.
+
 Parallelizing One Operation
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Set ``TF_INTRA_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used
-to parallelize the execution of one operation. If left unspecified, this value defaults to ``0`` which means TensorFlow should
+to parallelize the execution of one operation. For example, operations like ``tf.matmul()`` and ``tf.reduce_sum`` can be executed
+on multiple threads running parallely. The default value for this variable is ``0`` which means TensorFlow should
 pick an appropriate value depending on the system configuration.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]: what does "an appropriate value" actually mean? That it'll run on the maximum amount of possible threads? If yes, why even set this variable to anything?

File: docs/api/tensorflow_usage.rst [hunk 3]
```diff
@@ -14,19 +14,32 @@ All configuration options are specified using environment variables as shown in
...
 
 Parallelizing Multiple Operations
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Set ``TF_INTER_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used
-to parallelize the execution of multiple **non-blocking** operations. If left unspecified, this value defaults to ``0``
-which means TensorFlow should pick an appropriate value depending on the system configuration.
+to parallelize the execution of multiple **non-blocking** operations. These would included operations which do not have a
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
to parallelize the execution of multiple **non-blocking** operations. These would include operations that do not have a
```

File: docs/api/tensorflow_usage.rst [hunk 4]
```diff
@@ -14,19 +14,32 @@ All configuration options are specified using environment variables as shown in
...
+directed path between them in the tensorflow graph. The default value for this variable is ``0`` which means TensorFlow
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]: directed or direct?

File: docs/api/tensorflow_usage.rst [hunk 5]
```diff
@@ -14,19 +14,32 @@ All configuration options are specified using environment variables as shown in
...
+should pick an appropriate value depending on the system configuration.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]: what does appropriate value mean here as well?

File: docs/api/tensorflow_usage.rst [hunk 6]
```diff
@@ -35,18 +48,16 @@ Limiting GPU Memory Growth
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 TensorFlow by default blocks all the available GPU memory for the running process. This can be limiting if you are running
-multiple TensorFlow processes and want to distribute memory across them. To prevent this,
-set the environment variable ``TF_FORCE_GPU_ALLOW_GROWTH`` to ``True``.
+multiple TensorFlow processes and want to distribute memory across them. To prevent Rasa OS from blocking the
+complete available GPU memory, set the environment variable ``TF_FORCE_GPU_ALLOW_GROWTH`` to ``True``.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
multiple TensorFlow processes and want to distribute memory across them. To prevent Rasa Open Source from blocking all
of the available GPU memory, set the environment variable ``TF_FORCE_GPU_ALLOW_GROWTH`` to ``True``.
```

File: docs/api/tensorflow_usage.rst [hunk 7]
```diff
@@ -35,18 +48,16 @@ Limiting GPU Memory Growth
...
 
 Restricting Absolute GPU Memory Available
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Often, a developer wants to limit the absolute amount of GPU memory that can be used by a process.
+Often, a developer wants to limit the absolute amount of GPU memory that can be used by a Rasa OS process.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
You may want to limit the absolute amount of GPU memory that can be used by a Rasa Open Source process.
```

File: docs/api/tensorflow_usage.rst [hunk 8]
```diff
@@ -35,18 +48,16 @@ Limiting GPU Memory Growth
...
 
 For example, you may have two visible GPUs(``GPU:0`` and ``GPU:1``) and you want to allocate 1024 MB from the first GPU
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
For example, say you have two visible GPUs(``GPU:0`` and ``GPU:1``) and you want to allocate 1024 MB from the first GPU
```

File: docs/api/tensorflow_usage.rst [hunk 9]
```diff
@@ -35,18 +48,16 @@ Limiting GPU Memory Growth
...
-and 2048 MB from the second GPU.
-You can do so by setting an environment variable as ``TF_GPU_MEMORY_ALLOC="0:1024, 1:2048"``.
+and 2048 MB from the second GPU. You can do so by setting an environment variable as ``TF_GPU_MEMORY_ALLOC="0:1024, 1:2048"``.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
and 2048 MB from the second GPU. You can do this by setting the environment variable  ``TF_GPU_MEMORY_ALLOC`` to ``"0:1024, 1:2048"``.
```

File: docs/api/tensorflow_usage.rst [hunk 10]
```diff
@@ -35,18 +48,16 @@ Limiting GPU Memory Growth
...
 
 Another scenario can be where you have access to 2 GPUs(``GPU:0`` and ``GPU:1``) but you would like to use only the second
-GPU.
-``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory available from GPU 1.
+GPU. ``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory available from GPU 1.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]: do we really think we need the second scenario? Or do we think it's self explanatory with the first?

File: docs/api/tensorflow_usage.rst (hunk 2 above)
Comment (dakshvar22) [reply docs/api/tensorflow_usage.rst]: @akelad TensorFlow also does not mention anything about it in the docs. Assuming it runs on maximum amount of possible threads, as a user this may be undesirable because that hogs my CPU compute capacity and I may want to be conservative in it for other processed.

File: docs/api/tensorflow_usage.rst (hunk 4 above)
Comment (dakshvar22) [reply docs/api/tensorflow_usage.rst]: @akelad Should be directed because a direct path is something like A->B between two nodes A and B.
By directed I mean, consider A->B->D and C->D , then there is not directed path between A,C and B,C

File: docs/api/tensorflow_usage.rst (hunk 5 above)
Comment (dakshvar22) [reply docs/api/tensorflow_usage.rst]: Same as above

File: docs/api/tensorflow_usage.rst (hunk 10 above)
Comment (dakshvar22) [reply docs/api/tensorflow_usage.rst]: Agreed, removed.

File: docs/api/tensorflow_usage.rst (hunk 2 above)
Comment (dakshvar22) [reply docs/api/tensorflow_usage.rst]: Okay, I looked at the implementation and default value of 0 makes tensorflow allocate 1 thread per CPU core. Updated this here.

File: docs/api/tensorflow_usage.rst (hunk 4 above)
Comment (dakshvar22) [reply docs/api/tensorflow_usage.rst]: Added a small explanation

File: docs/api/tensorflow_usage.rst [hunk 11]
```diff
@@ -15,27 +15,28 @@ Optimizing CPU Performance
...
 .. note::
-    It is advisable to configure these options only if you are an advanced tensorflow user and understand
-    how different operations are implemented under the hood.
+    Since these options affect how operations are carried out under the hood in TensorFlow, it is advisable to
+    understand the implementation of different components first and then configure these options. You can always leave
+    them to their default value otherwise.
 
-These options are useful if you have implemented a custom component as part of your pipeline. Based on the tensorflow operations
-that the component uses, you can leverage multi-core CPU parallelism by tuning these options.
+Depending on the TensorFlow operations an already existing or a custom component uses, you can leverage multi-core CPU
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
Depending on the TensorFlow operations a NLU component or Core policy uses, you can leverage multi-core CPU
```

File: docs/api/tensorflow_usage.rst [hunk 12]
```diff
@@ -15,27 +15,28 @@ Optimizing CPU Performance
...
 Parallelizing Multiple Operations
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Set ``TF_INTER_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used
-to parallelize the execution of multiple **non-blocking** operations. These would included operations which do not have a
-directed path between them in the tensorflow graph. The default value for this variable is ``0`` which means TensorFlow
-should pick an appropriate value depending on the system configuration.
+to parallelize the execution of multiple **non-blocking** operations. These would include operations that do not have a
+directed path between them in the TensorFlow graph. In other words, the computation of one operation does not affect the
+computation of other operation. The default value for this variable is ``0`` which means TensorFlow would allocate one thread per CPU core.
```
Comment (akelad) [inline docs/api/tensorflow_usage.rst]:
```suggestion
computation of the other operation. The default value for this variable is ``0`` which means TensorFlow would allocate one thread per CPU core.
```
